### PR TITLE
Add integer option to initialize to init using bitmap

### LINKED ISSUE #5401: Should be able to initialize from an integer
### What is the expected enhancement?

Should be able to do:

````
qc.initialize(int)
````
and have it add x gates where the bit values are one.

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #5401

### Details and comments

This PR adds the option of using an integer to initialize a circuit. The integer acts as a bitmap creating an X gate on the appropriate wire for each of the 1 bits. For example,
```
from qiskit import BasicAer, execute, QuantumCircuit, QuantumRegister
qr = QuantumRegister(5)
qc = QuantumCircuit(qr)
qc.initialize(5, [0, 1, 3])
qc.measure_all()

backend = BasicAer.get_backend('qasm_simulator')
print(execute(qc, backend).result().get_counts())
qc.draw('text')
```
```
{'01001': 1024}

        ┌────────────────┐ ░ ┌─┐            
 q32_0: ┤0               ├─░─┤M├────────────
        │                │ ░ └╥┘┌─┐         
 q32_1: ┤1               ├─░──╫─┤M├─────────
        │  initialize(5) │ ░  ║ └╥┘┌─┐      
 q32_2: ┤                ├─░──╫──╫─┤M├──────
        │                │ ░  ║  ║ └╥┘┌─┐   
 q32_3: ┤2               ├─░──╫──╫──╫─┤M├───
        └────────────────┘ ░  ║  ║  ║ └╥┘┌─┐
 q32_4: ───────────────────░──╫──╫──╫──╫─┤M├
                           ░  ║  ║  ║  ║ └╥┘
meas: 5/══════════════════════╩══╩══╩══╩══╩═
                              0  1  2  3  4 
```

### CODE REVIEW

File: test/python/circuit/test_initializer.py [hunk 1]
```diff
@@ -357,6 +357,20 @@ def test_from_labels(self):
...
+
+        qc = QuantumCircuit(6)
+        qc.initialize(53, range(6))
+        job = execute(qc, BasicAer.get_backend('statevector_simulator'))
+        result = job.result()
+        statevector = result.get_statevector()
+        fidelity = state_fidelity(statevector, desired_sv)
+        self.assertGreater(
+            fidelity, self._desired_fidelity,
+            "Initializer has low fidelity {:.2g}.".format(fidelity))
```
Comment (Cryoris) [inline test/python/circuit/test_initializer.py:372]: Can you use the `Statevector` to do this check? 🙂 
```suggestion
        actual_sv = Statevector.from_instruction(qc)
        self.assertTrue(desired_sv == actual_sv)
```

File: qiskit/extensions/quantum_initializer/initializer.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ class Initialize(Instruction):
     which is not unitary.
     """
 
-    def __init__(self, params):
+    def __init__(self, params, num_qubits=None):
```
Comment (Cryoris) [inline qiskit/extensions/quantum_initializer/initializer.py:45]: Can you add a comment and error message that this argument is only valid if `params` is an integer? To capture cases where users provide e.g. labels and the number of qubits.

File: qiskit/extensions/quantum_initializer/initializer.py [hunk 3]
```diff
@@ -99,6 +110,27 @@ def _define_fromlabel(self):
 
         return initialize_circuit
 
+    def _define_from_int(self):
+        q = QuantumRegister(self.num_qubits, 'q')
+        initialize_circuit = QuantumCircuit(q, name='init_def')
+
+        # Convert to int since QuantumCircuit converted to complex
+        # and make a bit string and reverse it
+        intstr = f'{int(np.real(self.params[0])):0b}'[::-1]
```
Comment (Cryoris) [inline qiskit/extensions/quantum_initializer/initializer.py:119]: This doesn't apply the reset gate on all qubits right now (but it should):
```
>>> qc = QuantumCircuit(3)
>>> qc.initialize(1, [0, 1, 2])
>>> qc.draw()
     ┌────────────────┐
q_0: ┤0               ├
     │                │
q_1: ┤1 initialize(1) ├
     │                │
q_2: ┤2               ├
     └────────────────┘
>>> qc.decompose().draw()
          ┌───┐
q_0: ─|0>─┤ X ├
          └───┘
q_1: ──────────

q_2: ──────────
```

This is likely fixed by expanding the bitstring onto all qubits:
```suggestion
        param = int(np.real(self.params[0]))
        intstr = bin(param)[2:].zfill(self.num_qubits)[::-1]
```

**Review by Cryoris — CHANGES_REQUESTED** [review]
Mostly looks good, I left a few comments 👍

File: test/python/circuit/test_initializer.py (hunk 1 above)
Comment (enavarro51) [reply test/python/circuit/test_initializer.py]: I also changed the `test_from_label ` above this one to use `Statevector`.

Done in https://github.com/Qiskit/qiskit-terra/pull/5418/commits/8a7f33f292af59659e661aff949c9e31a25ec79a

File: qiskit/extensions/quantum_initializer/initializer.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/extensions/quantum_initializer/initializer.py:45]: Right. I forgot to add the int and num_qubits options to the doc string. Also added an error if not integer `params `and `num_qubits ` not None.

Done in https://github.com/Qiskit/qiskit-terra/pull/5418/commits/8a7f33f292af59659e661aff949c9e31a25ec79a

File: qiskit/extensions/quantum_initializer/initializer.py (hunk 3 above)
Comment (enavarro51) [reply qiskit/extensions/quantum_initializer/initializer.py]: Ah, good catch. I originally had `intstr = f'{int(np.real(self.params[0])):0{self.num_qubits}b}'[::-1]`, but lost the num_qubits somehow.

Done in https://github.com/Qiskit/qiskit-terra/pull/5418/commits/8a7f33f292af59659e661aff949c9e31a25ec79a

File: qiskit/extensions/quantum_initializer/initializer.py [hunk 4]
```diff
@@ -45,14 +45,24 @@ class Initialize(Instruction):
...
+               qubit one to `|0>`.
+          * int: an integer that is used as a bitmap indicating which qubits to initialize
+               to `|1>`. Example: setting params to 5 would initialize qubit 0 and qubit 2
+               to `|1>` and qubit 1 to `|0>`.
+        num_qubits (int): This parameter is only used if params is an int. Indicates the total
+            number of qubits in the `initialize` call. Example: `initialize` covers 5 qubits
+            and params is 3. This allows qubits 0 and 1 to be initialized to `|1>` and the
+            remaining 3 qubits to be initialized to `|0>`.
         """
+        if not isinstance(params, int) and num_qubits:
```
Comment (Cryoris) [inline qiskit/extensions/quantum_initializer/initializer.py:63]: Using `None` as boolean created some nasty bugs in the past (e.g. when the signature changed and suddenly ints or booleans were passed as positional args), it's just a little overhead to add `is not None` but much safer 🙂 
```suggestion
        if not isinstance(params, int) and num_qubits is not None:
```

Comment (enavarro51) [reply qiskit/extensions/quantum_initializer/initializer.py]: Agreed. Done in https://github.com/Qiskit/qiskit-terra/pull/5418/commits/e918604354719e4a05b696ba09a586f6a390a037

**Review by Cryoris — APPROVED** [review]
LGTM 👍

**Comment by enavarro51** [comment]
Hi, @Cryoris. Just checking to see if there should be an automerge here?

**Pull request merged**